File: examples/mechanics/FEM/ImpactingBar/ImpactingBar.cpp
/* Siconos is a program dedicated to modeling, simulation and control
 * of non smooth dynamical systems.
 *
 * Copyright 2024 INRIA.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
  V. Acary

  A Bar bouncing on the ground
  Simulation with a Time-Stepping scheme.
*/

#include <SiconosKernel.hpp>
#include <SiconosMatrix.hpp>
#include <SiconosVector.hpp>
#include <chrono>
#include <concepts>

#include "UserDefinedParameter.hpp"

using namespace std;
namespace user =
    user_defined_ref;  // To choose the set of parameters used in the current simulation

// #define TS_PROJ
// #define TS_COMBINED

int main(int argc, char* argv[]) {
  try {
    // ================= Creation of the model =======================

    // -------------------------
    // --- Dynamical systems ---
    // -------------------------
    std::cout << "====> Model loading ...\n";

    auto ndof = user::nDof;
    double l = user::L / ndof;  // length of an element

    siconos::algebra::SiconosDenseMatrix massMatrix{ndof, ndof};
    massMatrix.setZero();
    siconos::algebra::SiconosDenseMatrix stiffnessMatrix{ndof, ndof};
    stiffnessMatrix.setZero();

    massMatrix(0, 0) = 1.0 / 3.0;
    massMatrix(0, 1) = 1.0 / 6.0;
    stiffnessMatrix(0, 0) = 1.0;
    stiffnessMatrix(0, 1) = -1.0;

    for (unsigned int i = 1; i < ndof - 1; i++) {
      massMatrix(i, i) = 2.0 / 3.0;
      massMatrix(i, i - 1) = 1.0 / 6.0;
      massMatrix(i, i + 1) = 1.0 / 6.0;

      stiffnessMatrix(i, i) = 2.0;
      stiffnessMatrix(i, i - 1) = -1.0;
      stiffnessMatrix(i, i + 1) = -1.0;
    }

    massMatrix(ndof - 1, ndof - 1) = 1.0 / 3.0;
    massMatrix(ndof - 1, ndof - 2) = 1.0 / 6.0;

    stiffnessMatrix(ndof - 1, ndof - 2) = -1.0;
    stiffnessMatrix(ndof - 1, ndof - 1) = 1.0;

    massMatrix *= user::rho * user::S * l;
    stiffnessMatrix *= user::E * user::S / l;

    // -- Initial positions and velocities --
    siconos::algebra::SiconosVector q0{ndof};
    q0.setConstant(user::position_init);
    siconos::algebra::SiconosVector v0{ndof};
    v0.setConstant(user::velocity_init);

    // -- The dynamical system --
    auto bar = std::make_shared<siconos::modeling::LagrangianLinearTIDS>(
        q0, v0, massMatrix, siconos::algebra::alias_t);

    // -- Set stiffness matrix (weight) --
    bar->setStiffnessMatrix(stiffnessMatrix, siconos::algebra::alias_t);

    // -- Set external forces (weight) --
    siconos::algebra::SiconosVector weight{ndof};
    weight.setZero();
    bar->setConstantFext(weight, siconos::algebra::alias_t);

    // --------------------
    // --- Interactions ---
    // --------------------

    // -- nslaw --
    double e = 0.0;

    // Interaction bar-floor
    //
    auto H = std::make_shared<siconos::algebra::SiconosDenseMatrix>(1, ndof);
    (*H)(0, 0) = 1.0;

    auto nslaw = std::make_shared<siconos::modeling::NewtonImpactNSL>(e);
    auto relation = std::make_shared<siconos::modeling::LagrangianLinearTIR>(*H);

    auto inter = std::make_shared<siconos::modeling::Interaction>(nslaw, relation);

    // -------------
    // --- Model ---
    // -------------
    auto impactingBar =
        std::make_shared<siconos::modeling::NonSmoothDynamicalSystem>(user::t0, user::T);

    // add the dynamical system in the non smooth dynamical system
    impactingBar->insertDynamicalSystem(bar);

    // link the interaction and the dynamical system
    impactingBar->link(inter, bar);

    // ------------------
    // --- Simulation ---
    // ------------------

    // -- (1) OneStepIntegrators --
#ifdef TS_PROJ
    auto OSI =
        std::make_shared<siconos::integrators::MoreauJeanDirectProjectionOSI>(user::theta);
    OSI->setDeactivateYPosThreshold(1e-05);
    OSI->setDeactivateYVelThreshold(0.0);
    OSI->setActivateYPosThreshold(1e-09);
    OSI->setActivateYVelThreshold(100.0);
#else
#ifdef TS_COMBINED
    auto OSI = std::make_shared<siconos::integrators::MoreauJeanCombinedProjectionOSI>(theta);
#else
    auto OSI = std::make_shared<siconos::integrators::MoreauJeanOSI>(user::theta, 0.0);
    // OSI->setConstraintActivationThreshold(1e-05);
#endif
#endif
    // -- (2) Time discretisation --
    auto t = std::make_shared<siconos::simulation::TimeDiscretisation>(user::t0, user::h);

    // -- (3) one step non smooth problem
    auto osnspb = std::make_shared<siconos::nonsmooth_formulations::LCP>();

    // -- (4) Simulation setup with (1) (2) (3)
#ifdef TS_PROJ
    auto position =
        std::make_shared<siconos::nonsmooth_formulations::MLCPProjectOnConstraints>();
    auto s = std::make_shared<siconos::simulation::TimeSteppingDirectProjection>(
        impactingBar, t, OSI, osnspb, position, 0);
    s->setProjectionMaxIteration(10);
    s->setConstraintTolUnilateral(1e-10);
    s->setConstraintTol(1e-10);

#else
#ifdef TS_COMBINED
    auto position =
        std::make_shared<siconos::nonsmooth_formulations::MLCPProjectOnConstraints>(
            SICONOS_MLCP_ENUM);
    auto s = std::make_shared<siconos::simulation::TimeSteppingCombinedProjection>(
        impactingBar, t, OSI, osnspb, position, 2);
    s->setProjectionMaxIteration(500);
    s->setConstraintTolUnilateral(1e-10);
    s->setConstraintTol(1e-10);
#else
    auto s = std::make_shared<siconos::simulation::TimeStepping>(impactingBar, t, OSI, osnspb);
#endif
#endif

    // =========================== End of model definition ===========================

    // ================================= Computation =================================

    // --- Simulation initialization ---

    int N = floor((user::T - user::t0) / user::h);  // Number of time steps

    // --- Get the values to be plotted ---
    // -> saved in a matrix dataPlot
    unsigned int outputSize = 12;
    siconos::algebra::SiconosDenseMatrix dataPlot(N, outputSize);

    auto q = bar->q_read();
    auto v = bar->velocity_read();
    auto p = bar->p_read(1);
    auto lambda = inter->lambda(1);

    dataPlot(0, 0) = impactingBar->t0();
    dataPlot(0, 1) = q(0);
    dataPlot(0, 2) = v(0);
    dataPlot(0, 3) = p(0);
    dataPlot(0, 4) = (*lambda)(0);
    dataPlot(0, 7) = q(ndof - 1);
    dataPlot(0, 8) = v(ndof - 1);
    dataPlot(0, 9) = q((ndof) / 2);
    dataPlot(0, 10) = v((ndof) / 2);

    siconos::algebra::SiconosVector tmp{ndof};
    tmp = stiffnessMatrix * q;
    double potentialEnergy = q.dot(tmp);
    tmp = massMatrix * v;
    double kineticEnergy = v.dot(tmp);
    double impactEnergy = 0.0;

    dataPlot(0, 5) = potentialEnergy;
    dataPlot(0, 6) = kineticEnergy;
    dataPlot(0, 11) = impactEnergy;

    //    std::cout <<"potentialEnergy ="<<potentialEnergy << std::endl;
    //    std::cout <<"kineticEnergy ="<<kineticEnergy << std::endl;

    // --- Time loop ---
    cout << "====> Start computation ... \n\n";
    // ==== Simulation loop - Writing without explicit event handling =====
    int k = 1;

    auto start = std::chrono::system_clock::now();
    //    while (s->nextTime() < T)
    while (k < N) {
      s->advanceToEvent();

      //       std::cout << "position"  << std::endl;
      //       siconos::algebra::print(*q);
      //       std::cout << "velocity"  << std::endl;
      //       siconos::algebra::print(*v);

      // --- Get values to be plotted ---
      dataPlot(k, 0) = s->nextTime();
      dataPlot(k, 1) = q(0);
      dataPlot(k, 2) = v(0);
      dataPlot(k, 3) = p(0) / user::h;
      dataPlot(k, 4) = (*lambda)(0);
      dataPlot(k, 7) = q(ndof - 1);
      dataPlot(k, 8) = v(ndof - 1);
      dataPlot(k, 9) = q((ndof) / 2);
      dataPlot(k, 10) = v((ndof) / 2);

      tmp = stiffnessMatrix * q;
      potentialEnergy = q.dot(tmp);
      tmp = massMatrix * v;
      kineticEnergy = v.dot(tmp);
      dataPlot(k, 5) = potentialEnergy;
      dataPlot(k, 6) = kineticEnergy;

      double v_p_theta = user::theta * v(0) + (1 - user::theta) * dataPlot(k - 1, 2);
      impactEnergy = v_p_theta * (*lambda)(0);
      dataPlot(k, 11) = impactEnergy;

      //      std::cout << "q" << std::endl;
      //       siconos::algebra::print(*q);

      //       std::cout <<"potentialEnergy ="<<potentialEnergy << std::endl;
      //       std::cout <<"kineticEnergy ="<<kineticEnergy << std::endl;

      s->processEvents();

      k++;
    }
    double totalImpactEnergy = 0.0;
    for (int p = 0; p < k; p++) {
      totalImpactEnergy = totalImpactEnergy + dataPlot(p, 11);
    }
    printf("totalImpactEnergy = %e", totalImpactEnergy);

    auto end = std::chrono::system_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
    std::cout << "\nEnd of computation - Number of iterations done: " << k - 1;
    std::cout << "\nComputation time : " << elapsed << " ms\n";

    // --- Output files ---
    std::cout << "====> Output file writing ...\n";
    siconos::algebra::io::write("ImpactingBar.dat", dataPlot, siconos::algebra::io::ASCII_OUT,
                                siconos::algebra::io::WriteType::nodim);
    double error = 0.0, eps = 1e-9;
    if ((error = siconos::algebra::io::compareRefFile(dataPlot, "ImpactingBar.ref", eps)) >=
        eps)
      return 1;

    return 0;

  }

  catch (...) {
    siconos::exception::process();
    return 1;
  }
}

File: examples/mechanics/FEM/ImpactingBar/ImpactingBar.py
import numpy as np
import siconos.modeling as sm
import siconos.integrators
import siconos.simulation
import siconos.nonsmooth_formulations
import matplotlib.pyplot as plt


# User-defined main parameters
nDof = 500  # degrees of freedom for the beam
t0 = 1e-8  # initial computation time
T = 0.0015  # final computation time
h = 2e-7  # time step
position_init = 0.00005  # initial position
velocity_init = -0.1  # initial velocity
epsilon = 0.5  # 1e-1
theta = 1 / 2.0 + epsilon  # theta for MoreauJeanOSI integrator
# theta = 1.0
E = 210e9  # young Modulus
S = 0.000314  # Beam Section 1 cm  for the diameter
# S=0.1
beam_length = 1.0  # length of the  beam
elem_length = beam_length / nDof  # length of an element
rho = 7800.0  # specific mass
# rho=1.0
g = 9.81  # Gravity
g = 0.0


# mass= SiconosMatrix(nDof,nDof,SPARSE,nDof)
# stiffness= SiconosMatrix(nDof,nDof,SPARSE,nDof)
mass = np.zeros((nDof, nDof), dtype=np.float64, order="F")
stiffness = np.zeros((nDof, nDof), dtype=np.float64, order="F")

stiffness[0, 0] = 1.0 * E * S / elem_length
stiffness[0, 1] = -1.0 * E * S / elem_length
mass[0, 0] = 1 / 3.0 * rho * S * elem_length
mass[0, 1] = 1 / 6.0 * rho * S * elem_length

for i in range(1, nDof - 1):
    stiffness[i, i] = 2.0 * E * S / elem_length
    stiffness[i, i - 1] = -1.0 * E * S / elem_length
    stiffness[i, i + 1] = -1.0 * E * S / elem_length
    mass[i, i] = 2 / 3.0 * rho * S * elem_length
    mass[i, i - 1] = 1 / 6.0 * rho * S * elem_length
    mass[i, i + 1] = 1 / 6.0 * rho * S * elem_length


stiffness[nDof - 1, nDof - 2] = -1.0 * E * S / elem_length
stiffness[nDof - 1, nDof - 1] = 1.0 * E * S / elem_length
mass[nDof - 1, nDof - 2] = 1 / 6.0 * rho * S * elem_length
mass[nDof - 1, nDof - 1] = 1 / 3.0 * rho * S * elem_length


q0 = np.full((nDof), position_init)
v0 = np.full((nDof), velocity_init)

bar = sm.LagrangianLinearTIDS(q0, v0, mass)
bar.setStiffnessMatrix(stiffness)
# bar.display()

weight = np.full((nDof), -g * rho * S / elem_length)
bar.setConstantFext(weight)

e = 0.0

H = np.zeros((1, nDof))
H[0, 0] = 1.0

nslaw = sm.NewtonImpactNSL(e)
relation = sm.LagrangianLinearTIR(H)
inter = sm.Interaction(nslaw, relation)

# -------------
# --- Model ---
# -------------
impactingBar = sm.NonSmoothDynamicalSystem(t0, T)

# add the dynamical system in the non smooth dynamical system
impactingBar.insertDynamicalSystem(bar)

# link the interaction and the dynamical system
impactingBar.link(inter, bar)


# ------------------
# --- Simulation ---
# ------------------

# -- (1) OneStepIntegrators --
OSI = siconos.integrators.MoreauJeanOSI(theta, 0.5)

# -- (2) Time discretisation --
t = siconos.simulation.TimeDiscretisation(t0, h)

# -- (3) one step non smooth problem
osnspb = siconos.nonsmooth_formulations.LCP()

s = siconos.simulation.TimeStepping(impactingBar, t, OSI, osnspb)

k = 0

N = int((T - t0) / h)
dataPlot = np.zeros((N + 1, 5))

q = bar.q()
v = bar.velocity()
p = bar.p(1)
lambda_ = inter.lambda_python(1)

# time loop
while s.hasNextEvent():
    s.computeOneStep()
    dataPlot[k, 0] = s.nextTime()
    # print("time=", dataPlot[k, 0])
    dataPlot[k, 1] = q[0]
    dataPlot[k, 2] = v[0]
    dataPlot[k, 3] = p[0] / h
    dataPlot[k, 4] = lambda_[0]

    k += 1
    s.nextStep()

dataPlot.resize(k, 5)

fig_size = [14, 14]
plt.rcParams["figure.figsize"] = fig_size

plt.subplot(411)
plt.title("position")
plt.plot(dataPlot[:, 0], dataPlot[:, 1])
plt.grid()
plt.subplot(412)
plt.title("velocity")
plt.plot(dataPlot[:, 0], dataPlot[:, 2])
plt.grid()
plt.subplot(413)
plt.plot(dataPlot[:, 0], dataPlot[:, 3])
plt.title("reaction")
plt.grid()
plt.subplot(414)
plt.plot(dataPlot[:, 0], dataPlot[:, 4])
plt.title("lambda")
plt.grid()


plt.show()
